pynear/include/BKTree.hpp

```diff
@@ -97,37 +98,40 @@ template <typename key_t, typename distance_t, typename metric> class BKTree {
             dist = dist_opt.value();
 
             if (dist <= threshold) {
+                indices.push_back(candidate->index);
                 distances.push_back(dist);
                 keys.push_back(candidate->key);
             }
 
             lower = dist - threshold;
             upper = dist + threshold;
-            for (auto const &[d, bknode] : candidate->leaves) {
+            for (auto [d, bknode] : candidate->leaves) {
                 if (lower <= d && d <= upper) {
                     candidates.push_back(bknode);
                 }
             }
         }
-        return std::make_tuple(distances, keys);
+        return std::make_tuple(indices, distances, keys);
     }
 
-    std::tuple<std::vector<std::vector<distance_t>>, std::vector<std::vector<key_t>>> find_batch(const std::vector<key_t> &keys,
-                                                                                                 distance_t threshold) {
+    std::tuple<std::vector<std::vector<index_t>>, std::vector<std::vector<distance_t>>, std::vector<std::vector<key_t>>>
+    find_batch(const std::vector<key_t> &keys, distance_t threshold) {
+        std::vector<std::vector<index_t>> indices_out(keys.size());
         std::vector<std::vector<distance_t>> distances_out(keys.size());
         std::vector<std::vector<key_t>> keys_out(keys.size());
 
 #if (ENABLE_OMP_PARALLEL)
 #pragma omp parallel for schedule(static, 1)
 #endif
         // i should be size_t, however msvc requires signed integral loop variables (except with -openmp:llvm)
-        for (int i = 0; i < keys.size(); ++i) {
-            auto const &[distances_res, keys_res] = find(keys[i], threshold);
+        for (int i = 0; i < static_cast<int>(keys.size()); ++i) {
+            auto &&[indices_res, distances_res, keys_res] = find(keys[i], threshold);
+            indices_out[i] = std::move(indices_res);
             distances_out[i] = std::move(distances_res);
             keys_out[i] = std::move(keys_res);
```

Comment:
I think you dont need move here as it will call the proper move assignment operator. Move has no effect here afaict.
Yes, we cannot move from const, so the std::move operator will return the same const ref if we call it with const ref.